File: tools/pythonpkg/tests/fast/test_transaction.py
import duckdb
import pandas as pd

class TestConnectionTransaction(object):
    def test_transaction(self, duckdb_cursor):
        con = duckdb.connect()
        con.execute('create table t (i integer)')
        con.execute ('insert into t values (1)')

        con.begin()
        con.execute ('insert into t values (1)')
        assert con.execute('select count (*) from t').fetchone()[0] == 2
        con.rollback()
        assert con.execute('select count (*) from t').fetchone()[0] == 1
        con.begin()
        con.execute ('insert into t values (1)')
        assert con.execute('select count (*) from t').fetchone()[0] == 2
        con.commit()
        assert con.execute('select count (*) from t').fetchone()[0] == 2

File: tools/juliapkg/test/test_transaction.jl
# test_transaction.jl

@testset "Test DBInterface.transaction" begin
    con = DBInterface.connect(DuckDB.DB, ":memory:")

    # throw an exception in DBInterface.transaction
    # this should cause a rollback to happen
    @test_throws DuckDB.QueryException DBInterface.transaction(con) do
        DBInterface.execute(con, "CREATE TABLE integers(i INTEGER)")
        return DBInterface.execute(con, "SELEC")
    end

    # verify that the table does not exist
    @test_throws DuckDB.QueryException DBInterface.execute(con, "SELECT * FROM integers")

    # no exception, this should work and be committed
    DBInterface.transaction(con) do
        return DBInterface.execute(con, "CREATE TABLE integers(i INTEGER)")
    end
    DBInterface.execute(con, "SELECT * FROM integers")

    DBInterface.close!(con)
end
